L1Trigger/TrackFindingTracklet/src/VMRouterCM.cc

```diff
@@ -0,0 +1,283 @@
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/Utilities/interface/Exception.h"
+
+using namespace std;
+using namespace trklet;
+
+VMRouterCM::VMRouterCM(string name, Settings const& settings, Globals* global, unsigned int iSector)
+    : ProcessBase(name, settings, global, iSector), vmrtable_(settings) {
+  layerdisk_ = initLayerDisk(4);
+
+  vmstubsMEPHI_.resize(1, nullptr);
+
+  overlapbits_ = 7;
+  nextrabits_ = overlapbits_ - (settings_.nbitsallstubs(layerdisk_) + settings_.nbitsvmme(layerdisk_));
+
+  vmrtable_.init(layerdisk_, getName());
+
+  nbitszfinebintable_ = settings_.vmrlutzbits(layerdisk_);
+  nbitsrfinebintable_ = settings_.vmrlutrbits(layerdisk_);
+
+  nvmmebins_ = settings_.NLONGVMBINS() * ((layerdisk_ >= 6) ? 2 : 1);  //number of long z/r bins in VM
+}
+
+void VMRouterCM::addOutput(MemoryBase* memory, string output) {
+  if (settings_.writetrace()) {
+    edm::LogVerbatim("Tracklet") << "In " << name_ << " adding output to " << memory->getName() << " to output "
+                                 << output;
+  }
+
+  if (output.substr(0, 10) == "allstubout") {
+    AllStubsMemory* tmp = dynamic_cast<AllStubsMemory*>(memory);
+    assert(tmp != nullptr);
+    char memtype = 0;
+    if (output.size() > 10) {
+      memtype = output[11];
+    }
+    allstubs_.push_back(pair<char, AllStubsMemory*>(memtype, tmp));
+    return;
+  }
+
+  if (output.substr(0, 9) == "vmstubout") {
+    unsigned int pos = 12;
+    int vmbin = memory->getName().substr(pos, 1)[0] - '0';
+    pos++;
+    if (pos < memory->getName().size()) {
+      if (memory->getName().substr(pos, 1)[0] != 'n') {
+        vmbin = vmbin * 10 + memory->getName().substr(pos, 1)[0] - '0';
+        pos++;
+      }
+    }
+    if (memory->getName().substr(3, 2) == "TE") {
+      VMStubsTEMemory* tmp = dynamic_cast<VMStubsTEMemory*>(memory);
+      int iseed = output[output.size() - 1] - '0';
+      assert(iseed >= 0);
+      assert(iseed < 8);
+
+      int seedindex = -1;
+      for (unsigned int k = 0; k < vmstubsTEPHI_.size(); k++) {
+        if (vmstubsTEPHI_[k].seednumber == (unsigned int)iseed) {
+          seedindex = k;
+        }
+      }
+      if (seedindex == -1) {
+        seedindex = vmstubsTEPHI_.size();
+        vector<VMStubsTEMemory*> vectmp;
+        VMStubsTEPHICM atmp(iseed, vectmp);
+        vmstubsTEPHI_.push_back(atmp);
+      }
+      tmp->resize(settings_.NLONGVMBINS() * settings_.nvmte(1, iseed));
+      vmstubsTEPHI_[seedindex].vmstubmem.push_back(tmp);
+
+    } else if (memory->getName().substr(3, 2) == "ME") {
+      VMStubsMEMemory* tmp = dynamic_cast<VMStubsMEMemory*>(memory);
+      assert(tmp != nullptr);
+      tmp->resize(nvmmebins_ * settings_.nvmme(layerdisk_));
+      assert(vmstubsMEPHI_[0] == nullptr);
+      vmstubsMEPHI_[0] = tmp;
+    } else {
+      throw cms::Exception("LogicError") << __FILE__ << " " << __LINE__ << " should never get here!";
```

Comment:
probably more helpful to print the actual `memory->getName()` value to indicate why the exception occurs

Reply:
message updated